application/states.go

```diff
@@ -0,0 +1,228 @@
+	versionDetails VersionDetails) error {
+	fmt.Println("Associating")
+	err := models.ValidateVersion(combinedVersionUpdate)
+	if err != nil {
+		fmt.Println("Validation Failed")
+		fmt.Println(err)
+		return err
+	} else {
+		fmt.Println("Validation passed, continue")
+		return nil
+	}
+}
+func (g Associated) String() string {
+	return "associated"
+}
+
+type Published struct{}
+
+func (g Published) Enter(combinedVersionUpdate *models.Version, l *StateMachine, smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	fmt.Println("Publishing")
+
+	// This needs to do the validation on required fields etc.
+	errModel := models.ValidateVersion(combinedVersionUpdate)
+	if errModel != nil {
+		fmt.Println("Validation Failed")
+		fmt.Println(errModel)
+		return errModel
+	} else {
+		fmt.Println("Validation passed, continue")
+		eTag := headers.IfMatchAnyETag
+		if currentVersion.ETag != "" {
+			eTag = currentVersion.ETag
+		}
+
+		lockID, error := smDS.dataStore.Backend.AcquireInstanceLock(ctx, currentVersion.ID)
+		if error != nil {
+			return error
+		}
+		defer func() {
+			smDS.dataStore.Backend.UnlockInstance(ctx, lockID)
+		}()
+
+		if _, errVersion := smDS.dataStore.Backend.UpdateVersion(ctx, currentVersion, combinedVersionUpdate, eTag); errVersion != nil {
+			return errVersion
+		}
+
+		data := versionDetails.baseLogData()
+		log.Info(ctx, "attempting to publish version", data)
+		//errEdition := func() error {
+		editionDoc, err := smDS.dataStore.Backend.GetEdition(ctx, versionDetails.datasetID, versionDetails.edition, "")
+		if err != nil {
+			log.Error(ctx, "publishVersion: failed to find the edition we're trying to update", err, data)
+			return err
+		}
+
+		editionDoc.Next.State = models.PublishedState
+
+		if err := editionDoc.PublishLinks(ctx, versionUpdate.Links.Version); err != nil {
+			log.Error(ctx, "publishVersion: failed to update the edition links for the version we're trying to publish", err, data)
+			return err
+		}
+
+		editionDoc.Current = editionDoc.Next
+
+		if err := smDS.dataStore.Backend.UpsertEdition(ctx, versionDetails.datasetID, versionDetails.edition, editionDoc); err != nil {
+			log.Error(ctx, "publishVersion: failed to update edition during publishing", err, data)
+			return err
+		}
+
+		if err := smDS.dataStore.Backend.SetInstanceIsPublished(ctx, versionUpdate.ID); err != nil {
+			if user := dprequest.User(ctx); user != "" {
+				data[reqUser] = user
+			}
+			if caller := dprequest.Caller(ctx); caller != "" {
+				data[reqCaller] = caller
+			}
+			err := errors.WithMessage(err, "publishVersion: failed to set instance node is_published")
+			log.Error(ctx, "failed to publish instance version", err, data)
+			return err
+		}
```

Comment:
Consider either factoring this out into a separate function like `setCMDInstanceToPublished(...)` (recommended) and/or a code comment here to clarify that this is CMD only and is related to the Graph DB would be helpful as I found it confusing that we would be updating an instance record to published when we already changed the state when calling `UpdateVersion(...)` at the top of this function.

Furthermore, should this be wrapped in a conditional so that it only gets called for CMD?

Reply:
This has now been amended as suggested